tortoise/tortoise_test.go

```diff
@@ -2139,6 +2139,7 @@ func TestSwitchMode(t *testing.T) {
 		tortoise1 := tortoiseFromSimState(
 			s.GetState(0), WithConfig(cfg), WithLogger(logtest.New(t)),
 		)
+		require.NoError(t, tortoise1.WaitReady(ctx))
```

Comment:
While it fixes the test, I don't think it is a proper fix for the problem. The `Tortoise` still exposes a data-race-prone API and it should be fixed to prevent invalid use (calling `Tortoise::TallyVotes()` while the background recovery work is ongoing).

IMO, either `Tortoise::New()` should block until the recovery is done or the responsibility of making sure that the tortoise is ready should be moved to the caller of the `Tortoise::TallyVotes()` (like you did in the test). But then `Tortoise::TallyVotes()` should panic or return an error if Tortoise is not ready. I think panic is better as this is a bug in code logic. And panic is much better than UB caused by a data race.

Reply:
that makes sense in general, if you want please reopen a bug once this is merged